modules/object_table_ml.js

```diff
@@ -0,0 +1,163 @@
+const common = require("./common.js");
+
+/**
+ * A generic object table ML pipeline. 
+ * It incrementally performs an ML operation on new rows from the source table 
+ * and merges to the target table until no new row is detected or runs longer 
+ * than the specific duration.
+ * A row from the source table is considered as new if the `unique_key` (default to "uri") 
+ * of a row is absent in the target table, or if the `updated_column` (default to "updated") 
+ * column is newer than the largest value in the target table.
+ * 
+ * @param {Resolvable} source_table represents the source object table
+ * @param {String | Function} source either a query string or a Contextable function to produce the query on the source data
+ * @param {String} target_table the name of the table to store the final result
+ * @param {String} accept_filter a SQL expression for finding rows that contains retryable error
+ * @param {Number} batch_size number of rows to process in each SQL job. Rows in the object table will be 
+ *                 processed in batches according to the batch size. Default batch size is 500
+ * @param {String} unique_key the primary key in the target table for incremental update. Default value is "uri".
+ * @param {String} updated_column the column that carries the last updated timestamp of an object in the object 
+ *                 table. Default value is "updated"
+ * @param {Number} batch_duration_secs the number of seconds to pass before breaking the batching loop if it 
+ *                 hasn't been finished before within this duration. Default value is 22 hours
+ */
+const obj_table_ml = (source_table, source, target_table, accept_filter, {
+    batch_size = 500,
+    unique_key = "uri",
+    updated_column = "updated",
+    batch_duration_secs = 22 * 60 * 60,
+} = {}) => {
+    let source_func = (source instanceof Function) ? source : () => source;
+    let limit_clause = `LIMIT ${batch_size}`;
+
+    // Initialize by creating the target table with a small limit to avoid timeout
+    operate(`init_${target_table}`)
+        .queries((ctx) =>
+            `CREATE TABLE IF NOT EXISTS ${ctx.resolve(target_table)} AS ${source_func(ctx)} WHERE ${accept_filter} LIMIT 10`);
```

Comment:
Creating a table by a Dataform action, that isn't made by the Dataform action itself, breaks the Dataform paradigm a bit.

IIUC, this shouldn't actually be needed anyways - the preOps below is what references `target_table`, but that only happens on the incremental run due to the `ctx.when(ctx.incremental()`, when the table should exist. So I would be surprised if this function didn't work without this `operate(` removed entirely.

Reply:
This is because I want the function contract be "Give me a source table and a ML function, it will guarantee to process all rows with the ML function".

If I just rely on the `publish` with incremental type, it would require the script to be triggered twice at least, with the first run to create the table, and the second one to have a repeat loop and merge statement.

I admit that it is a bit hacky to have this init operation, but just to make this package easier to consume without a strict requirement on scheduling.

I also tried using two `publish` call, but seems like dataform disallow to have two publish calls on the same table.